File: frontend/src/lib/multiValue.ts
/**
 * Çok değerli hücre sözleşmesi (G124) — backend `services/multi_value.py` ikizi.
 *
 * Tıbbi beşli (tıbbi süreç / olay / iddia edilen kusur / hastada oluşan zarar /
 * uygulanan yöntem) tek metin kolonunda " ; " ile birleşik saklanır. Ayraç
 * noktalı virgüldür; virgül ayraç DEĞİLDİR (değerler virgül içerebilir).
 */
export const MULTI_SEPARATOR = " ; ";

const normalizeKey = (s: string): string => s.trim().toLocaleLowerCase("tr-TR");

/** Hücreyi parçalara ayırır: kırpar, boş ve mükerrer parçayı düşürür (ilk sıra korunur). */
export function splitValues(value: unknown): string[] {
    if (value === null || value === undefined) return [];
    const seen = new Set<string>();
    const out: string[] = [];
    for (const raw of String(value).split(/[;\r\n]+/)) {
        const part = raw.replace(/\s+/g, " ").trim();
        const key = normalizeKey(part);
        if (!part || seen.has(key)) continue;
        seen.add(key);
        out.push(part);
    }
    return out;
}

/** Parçaları kanonik ayraçla birleştirir; boş → "" (panel "" → null → alan silinir). */
export function joinValues(values: string[]): string {
    return values.map(v => v.replace(/\s+/g, " ").trim()).filter(Boolean).join(MULTI_SEPARATOR);
}

File: backend/services/multi_value.py
"""Çok değerli hücre sözleşmesi (G124) — TEK ayraç, TEK yazım.

Teslim paketinin beş tasnif sütunu (Tıbbi Süreç · Tıbbi Olay · İddia Edilen
Kusur · Hastada Oluşan Zarar · Uygulanan Yöntem) bir hücrede birden çok
değer taşır ("Komplikasyon Yönetimi ; Takip Eksikliği"). Kartta aynı metin
kolonunda saklanır; ayraç noktalı virgüldür — virgül ayraç DEĞİLDİR
(değerlerin kendisi virgül içerebilir: "Kaynamama / Yanlış Kaynama").

Bu modül ayırma/birleştirmenin tek yeridir: aktarım (`hukdok_aktarim`),
takip paneli yazma kapısı (`case_manager.validated_multi_list_value`) ve
havuz seed'i (`scripts/deger_havuzu_seed.py`) buradan okur; frontend
karşılığı `frontend/src/lib/multiValue.ts` (aynı ayraç).
"""
from __future__ import annotations

import re
from typing import Any, Iterable, List, Optional

SEPARATOR = " ; "
_SPLIT = re.compile(r"[;\r\n]+")


def split_values(value: Any) -> List[str]:
    """Hücreyi parçalara ayırır: boşluk kırpılır, boş parça ve mükerrer düşer
    (ilk görülen sıra korunur). None/boş → []."""
    if value is None:
        return []
    gorulen: List[str] = []
    for parca in _SPLIT.split(str(value)):
        temiz = " ".join(parca.split())
        if temiz and temiz.casefold() not in {g.casefold() for g in gorulen}:
            gorulen.append(temiz)
    return gorulen


def join_values(values: Iterable[str]) -> Optional[str]:
    """Parçaları kanonik ayraçla birleştirir; boş → None (alan temizlenir)."""
    parcalar = [p for p in (" ".join(str(v).split()) for v in values) if p]
    return SEPARATOR.join(parcalar) if parcalar else None
